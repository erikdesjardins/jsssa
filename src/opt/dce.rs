use crate::ir;
use crate::ir::fold::Folder;

#[derive(Default)]
pub struct Dce {
    dropping_after_jump: bool,
}

impl Folder for Dce {
    type Output = Option<ir::Stmt>;

    fn wrap_scope<R>(&mut self, enter: impl FnOnce(&mut Self) -> R) -> R {
        let r = enter(self);
        // stop dropping when we leave the current scope
        self.dropping_after_jump = false;
        r
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        if self.dropping_after_jump {
            return None;
        }

        match stmt {
            ir::Stmt::Expr {
                ref target,
                ref expr,
            } if target.used().is_never() => match expr {
                ir::Expr::Function { .. }
                | ir::Expr::Bool { .. }
                | ir::Expr::Number { .. }
                | ir::Expr::String { .. }
                | ir::Expr::Null
                | ir::Expr::Undefined
                | ir::Expr::This
                | ir::Expr::Read { .. }
                | ir::Expr::ReadMutable { .. }
                | ir::Expr::Array { .. }
                | ir::Expr::Object { .. }
                | ir::Expr::RegExp { .. }
                | ir::Expr::Unary { .. }
                | ir::Expr::Binary { .. }
                | ir::Expr::CurrentFunction
                | ir::Expr::Argument { .. } => None,
                ir::Expr::ReadGlobal { .. }
                | ir::Expr::ReadMember { .. }
                | ir::Expr::Delete { .. }
                | ir::Expr::Yield { .. }
                | ir::Expr::Await { .. }
                | ir::Expr::Call { .. } => Some(stmt),
            },
            ir::Stmt::DeclareMutable { ref target, val: _ } if target.used().is_never() => None,
            ir::Stmt::Return { .. }
            | ir::Stmt::Throw { .. }
            | ir::Stmt::Break
            | ir::Stmt::Continue => {
                self.dropping_after_jump = true;
                Some(stmt)
            }
            ir::Stmt::IfElse {
                cond: _,
                ref cons,
                ref alt,
            } if cons.0.is_empty() && alt.0.is_empty() => None,
            ir::Stmt::Try {
                ref body,
                catch: _,
                ref finally,
            } if body.0.is_empty() && finally.0.is_empty() => None,
            _ => Some(stmt),
        }
    }
}
