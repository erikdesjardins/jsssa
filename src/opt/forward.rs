use std::collections::HashMap;
use std::iter;

use crate::ir;
use crate::ir::traverse::{visit_with, Folder, ScopeTy};

/// Forward `ir::Expr::Read` to the source SSA ref.
///
/// Does not profit from multiple passes.
/// Does not profit from DCE running first; may create opportunities for DCE.
#[derive(Default)]
pub struct Reads {
    ssa_remappings: HashMap<ir::WeakRef<ir::Ssa>, ir::Ref<ir::Ssa>>,
}

impl Reads {
    fn remap(&self, ref_: ir::Ref<ir::Ssa>) -> ir::Ref<ir::Ssa> {
        match self.ssa_remappings.get(&ref_.weak()) {
            Some(orig_ref) => orig_ref.clone(),
            None => ref_,
        }
    }
}

impl Folder for Reads {
    type Output = iter::Once<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            visit_with(&block, |stmt: &ir::Stmt| match stmt {
                ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Read { source },
                } => match self.ssa_remappings.get(&source.weak()) {
                    Some(orig_ref) => {
                        self.ssa_remappings.insert(target.weak(), orig_ref.clone());
                    }
                    None => {
                        self.ssa_remappings.insert(target.weak(), source.clone());
                    }
                },
                _ => {}
            });
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        iter::once(match stmt {
            ir::Stmt::Expr { target, expr } => ir::Stmt::Expr {
                target,
                expr: match expr {
                    ir::Expr::Read { source } => ir::Expr::Read {
                        source: self.remap(source),
                    },
                    ir::Expr::ReadMember { obj, prop } => ir::Expr::ReadMember {
                        obj: self.remap(obj),
                        prop: self.remap(prop),
                    },
                    ir::Expr::Array { elems } => ir::Expr::Array {
                        elems: elems
                            .into_iter()
                            .map(|opt| opt.map(|(kind, ele)| (kind, self.remap(ele))))
                            .collect(),
                    },
                    ir::Expr::Object { props } => ir::Expr::Object {
                        props: props
                            .into_iter()
                            .map(|(kind, obj, prop)| (kind, self.remap(obj), self.remap(prop)))
                            .collect(),
                    },
                    ir::Expr::Unary { op, val } => ir::Expr::Unary {
                        op,
                        val: self.remap(val),
                    },
                    ir::Expr::Binary { op, left, right } => ir::Expr::Binary {
                        op,
                        left: self.remap(left),
                        right: self.remap(right),
                    },
                    ir::Expr::Delete { obj, prop } => ir::Expr::Delete {
                        obj: self.remap(obj),
                        prop: self.remap(prop),
                    },
                    ir::Expr::Yield { kind, val } => ir::Expr::Yield {
                        kind,
                        val: self.remap(val),
                    },
                    ir::Expr::Await { val } => ir::Expr::Await {
                        val: self.remap(val),
                    },
                    ir::Expr::Call { kind, func, args } => ir::Expr::Call {
                        kind,
                        func: self.remap(func),
                        args: args
                            .into_iter()
                            .map(|(kind, arg)| (kind, self.remap(arg)))
                            .collect(),
                    },
                    ir::Expr::Bool { .. }
                    | ir::Expr::Number { .. }
                    | ir::Expr::String { .. }
                    | ir::Expr::Null
                    | ir::Expr::Undefined
                    | ir::Expr::This
                    | ir::Expr::ReadMutable { .. }
                    | ir::Expr::ReadGlobal { .. }
                    | ir::Expr::RegExp { .. }
                    | ir::Expr::Function { .. }
                    | ir::Expr::CurrentFunction
                    | ir::Expr::Argument { .. } => expr,
                },
            },
            ir::Stmt::DeclareMutable { target, val } => ir::Stmt::DeclareMutable {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteMutable { target, val } => ir::Stmt::WriteMutable {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteGlobal { target, val } => ir::Stmt::WriteGlobal {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteMember { obj, prop, val } => ir::Stmt::WriteMember {
                obj: self.remap(obj),
                prop: self.remap(prop),
                val: self.remap(val),
            },
            ir::Stmt::Return { val } => ir::Stmt::Return {
                val: self.remap(val),
            },
            ir::Stmt::Throw { val } => ir::Stmt::Throw {
                val: self.remap(val),
            },
            ir::Stmt::ForEach { kind, init, body } => ir::Stmt::ForEach {
                kind,
                init: self.remap(init),
                body,
            },
            ir::Stmt::IfElse { cond, cons, alt } => ir::Stmt::IfElse {
                cond: self.remap(cond),
                cons,
                alt,
            },
            ir::Stmt::Switch { discr, body } => ir::Stmt::Switch {
                discr: self.remap(discr),
                body,
            },
            ir::Stmt::Break { .. }
            | ir::Stmt::Continue { .. }
            | ir::Stmt::Debugger
            | ir::Stmt::Label { .. }
            | ir::Stmt::Loop { .. }
            | ir::Stmt::SwitchCase { .. }
            | ir::Stmt::Try { .. } => stmt,
        })
    }
}
