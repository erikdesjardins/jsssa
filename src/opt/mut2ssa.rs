use std::collections::{HashMap, HashSet};
use std::mem;

use crate::ir;
use crate::ir::traverse::{Folder, RunVisitor, ScopeTy, Visitor};

/// Converts read-only mutable vars to SSA, and removes write-only mutable vars.
///
/// Does not profit from multiple passes.
/// May profit from DCE running first; may create opportunities for DCE.
/// May create opportunities for read forwarding.
#[derive(Default)]
pub struct Mut2Ssa {
    mut_vars_to_replace: HashMap<ir::WeakRef<ir::Mut>, What>,
}

enum What {
    Convert(ir::Ref<ir::Ssa>),
    Remove,
}

#[derive(Default)]
struct CollectMutOpInfo<'a> {
    mut_ops: HashMap<&'a ir::Ref<ir::Mut>, State>,
    reads_in_scope: HashSet<&'a ir::Ref<ir::Mut>>,
    declared_at_toplevel_of_switch: HashSet<&'a ir::Ref<ir::Mut>>,
    at_toplevel_of_switch: bool,
    about_to_enter_switch: bool,
}

#[derive(PartialEq)]
enum State {
    ReadOnly { frozen: bool },
    WriteOnly,
    Invalid,
}

impl<'a> Visitor<'a> for CollectMutOpInfo<'a> {
    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: &'a ir::Block,
        enter: impl FnOnce(&mut Self, &'a ir::Block) -> R,
    ) -> R {
        match ty {
            ScopeTy::Function => {
                // functions are analyzed separately, since they can read ssa refs
                // before they are lexically declared
                let mut inner = Self::default();
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                let r = enter(&mut inner, block);
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                r
            }
            ScopeTy::Normal | ScopeTy::Toplevel | ScopeTy::Nonlinear => {
                let mut inner = Self::default();
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                mem::swap(&mut inner.reads_in_scope, &mut self.reads_in_scope);
                inner.at_toplevel_of_switch = self.about_to_enter_switch;
                let r = enter(&mut inner, block);
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                mem::swap(&mut inner.reads_in_scope, &mut self.reads_in_scope);
                self.about_to_enter_switch = false;
                r
            }
        }
    }

    fn visit(&mut self, stmt: &'a ir::Stmt) {
        match stmt {
            ir::Stmt::Expr {
                target: _,
                expr: ir::Expr::ReadMutable { source },
            } => {
                let our_state = State::ReadOnly { frozen: false };
                self.mut_ops
                    .entry(source)
                    .and_modify(|state| {
                        if state != &our_state {
                            *state = State::Invalid;
                        }
                    })
                    .or_insert(our_state);
                self.reads_in_scope.insert(source);
            }
            ir::Stmt::WriteMutable { target, val: _ } => {
                let our_state = State::WriteOnly;
                self.mut_ops
                    .entry(target)
                    .and_modify(|state| {
                        if state != &our_state {
                            *state = State::Invalid;
                        }
                    })
                    .or_insert(our_state);
            }
            ir::Stmt::DeclareMutable { target, val: _ } => {
                if self.reads_in_scope.contains(&target) {
                    // read before declaration
                    self.mut_ops.insert(target, State::Invalid);
                }
                if self.at_toplevel_of_switch {
                    self.declared_at_toplevel_of_switch.insert(target);
                }
            }
            ir::Stmt::Switch { .. } => {
                self.about_to_enter_switch = true;
            }
            ir::Stmt::SwitchCase { .. } => {
                for ref_ in self.declared_at_toplevel_of_switch.drain() {
                    match self.mut_ops.entry(ref_).or_insert(State::WriteOnly) {
                        State::ReadOnly { frozen } => {
                            *frozen = true;
                        }
                        State::WriteOnly | State::Invalid => {
                            // future reads already disallowed
                        }
                    }
                }
            }
            _ => {}
        }
    }
}

impl Folder for Mut2Ssa {
    type Output = Option<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            let mut collector = CollectMutOpInfo::default();
            collector.run_visitor(&block);
            self.mut_vars_to_replace = collector
                .mut_ops
                .into_iter()
                .flat_map(|(ref_, saw)| match saw {
                    State::ReadOnly { frozen: _ } => {
                        Some((ref_.weak(), What::Convert(ir::Ref::new(ref_.name_hint()))))
                    }
                    State::WriteOnly => Some((ref_.weak(), What::Remove)),
                    State::Invalid => None,
                })
                .collect();
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        match stmt {
            ir::Stmt::Expr {
                target,
                expr: ir::Expr::ReadMutable { source },
            } => match self.mut_vars_to_replace.get(&source.weak()) {
                Some(What::Convert(ssa_ref)) => Some(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Read {
                        source: ssa_ref.clone(),
                    },
                }),
                Some(What::Remove) => unreachable!("removing mut read: {:?}", source),
                None => Some(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::ReadMutable { source },
                }),
            },
            ir::Stmt::DeclareMutable { target, val } => {
                match self.mut_vars_to_replace.get(&target.weak()) {
                    Some(What::Convert(ssa_ref)) => Some(ir::Stmt::Expr {
                        target: ssa_ref.clone(),
                        expr: ir::Expr::Read { source: val },
                    }),
                    Some(What::Remove) => None,
                    None => Some(ir::Stmt::DeclareMutable { target, val }),
                }
            }
            ir::Stmt::WriteMutable { target, val } => {
                match self.mut_vars_to_replace.get(&target.weak()) {
                    Some(What::Convert(_)) => unreachable!("converting mut write: {:?}", target),
                    Some(What::Remove) => None,
                    None => Some(ir::Stmt::WriteMutable { target, val }),
                }
            }
            _ => Some(stmt),
        }
    }
}
