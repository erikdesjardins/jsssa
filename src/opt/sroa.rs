use std::collections::{HashMap, HashSet};
use std::iter;

use crate::collections::ZeroOneMany::{self, Many, One};
use crate::ir;
use crate::ir::traverse::{Folder, RunVisitor, ScopeTy, Visitor};

/// Scalar replacement of aggregates (objects).
///
/// May profit from multiple passes.
/// Does not profit from DCE running first; may create opportunities for DCE.
/// May create opportunities for mut-to-ssa downleveling.
/// May create opportunities for read forwarding.
#[derive(Debug, Default)]
pub struct Replace {
    objects_to_replace: HashMap<ir::WeakRef<ir::Ssa>, HashMap<String, ir::Ref<ir::Mut>>>,
    known_strings: HashMap<ir::WeakRef<ir::Ssa>, String>,
}

#[derive(Debug, Default)]
struct CollectObjInfo<'a> {
    known_objs: HashMap<&'a ir::Ref<ir::Ssa>, State<'a>>,
    known_strings: HashMap<&'a ir::Ref<ir::Ssa>, &'a str>,
    last_use_was_safe: bool,
}

#[derive(Debug)]
enum State<'a> {
    NoObjYet(HashSet<&'a str>),
    HasProps(HashSet<&'a str>),
    Invalid,
}

fn is_safe_prop(prop: &str) -> bool {
    const OBJ_PROTO_PROPS: [&str; 12] = [
        "constructor",
        "hasOwnProperty",
        "isPrototypeOf",
        "propertyIsEnumerable",
        "toLocaleString",
        "toString",
        "valueOf",
        "__defineGetter__",
        "__defineSetter__",
        "__lookupGetter__",
        "__lookupSetter__",
        "__proto__",
    ];
    !OBJ_PROTO_PROPS.contains(&prop)
}

impl<'a> CollectObjInfo<'a> {
    fn declare_simple_object(
        &mut self,
        obj: &'a ir::Ref<ir::Ssa>,
        props: impl IntoIterator<Item = &'a ir::Ref<ir::Ssa>>,
    ) {
        let known_safe_props = props
            .into_iter()
            .map(|prop| match self.known_strings.get(prop) {
                Some(prop) if is_safe_prop(prop) => Ok(*prop),
                _ => Err(()),
            })
            .collect::<Result<_, _>>();
        match (self.known_objs.get_mut(obj), known_safe_props) {
            (None, Ok(props)) => {
                self.known_objs.insert(obj, State::HasProps(props));
            }
            (Some(State::NoObjYet(seen_props)), Ok(ref mut props)) => {
                let all_props = seen_props.drain().chain(props.drain()).collect();
                self.known_objs.insert(obj, State::HasProps(all_props));
            }
            (Some(State::HasProps(_)), _) => unreachable!("multiple ssa defns"),
            (Some(State::Invalid), _) => { /* already invalid */ }
            (_, Err(())) => {
                self.known_objs.insert(obj, State::Invalid);
            }
        }
    }

    fn access_prop(&mut self, obj: &'a ir::Ref<ir::Ssa>, prop: &'a ir::Ref<ir::Ssa>) {
        let known_safe_prop = match self.known_strings.get(prop) {
            Some(prop) if is_safe_prop(prop) => Some(*prop),
            _ => None,
        };
        match (self.known_objs.get_mut(obj), known_safe_prop) {
            (None, Some(prop)) => {
                self.known_objs
                    .insert(obj, State::NoObjYet(iter::once(prop).collect()));
            }
            (Some(State::NoObjYet(props)), Some(prop)) => {
                props.insert(prop);
            }
            (Some(State::HasProps(props)), Some(prop)) => {
                props.insert(prop);
            }
            (Some(State::Invalid), _) => { /* already invalid */ }
            (_, None) => {
                self.known_objs.insert(obj, State::Invalid);
            }
        }
    }
}

impl<'a> Visitor<'a> for CollectObjInfo<'a> {
    fn visit(&mut self, stmt: &'a ir::Stmt) {
        self.last_use_was_safe = false;

        match stmt {
            ir::Stmt::Expr { target, expr } => match expr {
                ir::Expr::String { value } => {
                    self.known_strings.insert(target, &value);
                }
                ir::Expr::Object { props } => {
                    let all_simple_props = props
                        .iter()
                        .map(|(kind, prop, _val)| match kind {
                            ir::PropKind::Simple => Ok(prop),
                            ir::PropKind::Get | ir::PropKind::Set => Err(()),
                        })
                        .collect::<Result<Vec<_>, _>>();
                    match all_simple_props {
                        Ok(simple_props) => {
                            self.declare_simple_object(target, simple_props);
                        }
                        Err(()) => {
                            self.known_objs.insert(target, State::Invalid);
                        }
                    }
                }
                ir::Expr::ReadMember { obj, prop } => {
                    self.last_use_was_safe = true;
                    self.access_prop(obj, prop);
                    self.known_objs.insert(prop, State::Invalid);
                }
                _ => {}
            },
            ir::Stmt::WriteMember { obj, prop, val } => {
                self.last_use_was_safe = true;
                self.access_prop(obj, prop);
                self.known_objs.insert(prop, State::Invalid);
                self.known_objs.insert(val, State::Invalid);
            }
            _ => {}
        }
    }

    fn visit_ref_use(&mut self, ref_: &'a ir::Ref<ir::Ssa>) {
        if !self.last_use_was_safe {
            self.known_objs.insert(ref_, State::Invalid);
        }
    }
}

impl Folder for Replace {
    type Output = ZeroOneMany<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            let mut collector = CollectObjInfo::default();
            collector.run_visitor(&block);
            self.objects_to_replace = collector
                .known_objs
                .into_iter()
                .filter_map(|(obj, state)| match state {
                    State::HasProps(props) => {
                        let prop_vars = props
                            .into_iter()
                            .map(|prop| (prop.to_string(), ir::Ref::new(prop)))
                            .collect();
                        Some((obj.weak(), prop_vars))
                    }
                    State::NoObjYet(_) | State::Invalid => None,
                })
                .collect();
            self.known_strings = collector
                .known_strings
                .into_iter()
                .map(|(ref_, str)| (ref_.weak(), str.to_string()))
                .collect();
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        match stmt {
            ir::Stmt::Expr {
                target,
                expr: ir::Expr::Object { props },
            } => match self.objects_to_replace.get(&target.weak()) {
                Some(prop_vars) => {
                    let mut prop_vars = prop_vars.clone();
                    let undef_ref = ir::Ref::new("_mis");
                    let undef = ir::Stmt::Expr {
                        target: undef_ref.clone(),
                        expr: ir::Expr::Undefined,
                    };
                    let props_with_values = props
                        .into_iter()
                        .map(|(kind, prop, val)| {
                            match (
                                kind,
                                self.known_strings
                                    .get(&prop.weak())
                                    .and_then(|name| prop_vars.remove(name)),
                            ) {
                                (ir::PropKind::Simple, Some(prop_var)) => {
                                    ir::Stmt::DeclareMutable {
                                        target: prop_var,
                                        val,
                                    }
                                }
                                (ir::PropKind::Get, _) | (ir::PropKind::Set, _) | (_, None) => {
                                    unreachable!("bad prop kind or unknown name")
                                }
                            }
                        })
                        .collect::<Vec<_>>();
                    let props_without_values =
                        prop_vars
                            .into_iter()
                            .map(|(_, prop_var)| ir::Stmt::DeclareMutable {
                                target: prop_var,
                                val: undef_ref.clone(),
                            });
                    Many(
                        iter::once(undef)
                            .chain(props_with_values)
                            .chain(props_without_values)
                            .collect(),
                    )
                }
                None => One(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Object { props },
                }),
            },
            ir::Stmt::Expr {
                target,
                expr: ir::Expr::ReadMember { obj, prop },
            } => match self.objects_to_replace.get(&obj.weak()) {
                Some(prop_vars) => {
                    // todo store a map from Ref<Ssa> prop refs to Ref<Mut> to avoid double lookup
                    let prop_var = match self
                        .known_strings
                        .get(&prop.weak())
                        .and_then(|name| prop_vars.get(name))
                    {
                        Some(prop_var) => prop_var.clone(),
                        None => unreachable!("unknown prop"),
                    };
                    One(ir::Stmt::Expr {
                        target,
                        expr: ir::Expr::ReadMutable { source: prop_var },
                    })
                }
                None => One(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::ReadMember { obj, prop },
                }),
            },
            ir::Stmt::WriteMember { obj, prop, val } => {
                match self.objects_to_replace.get(&obj.weak()) {
                    Some(prop_vars) => {
                        let prop_var = match self
                            .known_strings
                            .get(&prop.weak())
                            .and_then(|name| prop_vars.get(name))
                        {
                            Some(prop_var) => prop_var.clone(),
                            None => unreachable!("unknown prop"),
                        };
                        One(ir::Stmt::WriteMutable {
                            target: prop_var,
                            val,
                        })
                    }
                    None => One(ir::Stmt::WriteMember { obj, prop, val }),
                }
            }
            _ => One(stmt),
        }
    }
}
