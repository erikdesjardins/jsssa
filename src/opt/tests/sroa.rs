use crate::opt::dce;
use crate::opt::forward;
use crate::opt::mut2ssa;
use crate::opt::sroa;

macro_rules! passes {
    ( $cx:ident ) => {
        $cx.run::<mut2ssa::Mut2Ssa>("mut2ssa")
            .run::<forward::Reads>("forward-reads-redundancy")
            .converge::<dce::Dce>("dce-forwarded-reads")
            .run::<sroa::Replace>("scalar-replace")
    };
}

case!(
    basic,
    |cx| passes!(cx),
    r#"
    let something = { x: 1 };
    foo();
    g1 = something.x;
    something.x = g2;
"#
);

case!(
    invalid_unknown_prop,
    |cx| passes!(cx),
    r#"
    let something = { x: 1 };
    foo();
    something.prototype;
    g = something.x;
"#
);

case!(
    invalidate_escape,
    |cx| passes!(cx),
    r#"
    let something = { x: 1 };
    if (foo) {
        g = something;
    }
    g1 = something.x;
"#
);

case!(
    invalidate_escape_to_other_object,
    |cx| passes!(cx),
    r#"
    let something = { x: 1 };
    if (foo) {
        g = { foo: something };
    }
    g1 = something.x;
"#
);

case!(
    time_travel,
    |cx| passes!(cx),
    r#"
    g = function() {
        something.x;
    }
    let something = { x: 1 };
    something.x;
"#
);

case!(
    complex,
    |cx| passes!(cx),
    r#"
    let obj = { foo: 1, bar: 2, not_written: 2 };
    if (something) {
        obj.foo;
        obj.foo = 3;
        obj.bar;
        if (something2) {
            obj.bar = 4;
        }
        obj.foo;
        obj.not_written;
    }
    obj.foo;
    obj.bar;
    obj.not_written;
"#
);

case!(
    many_writes,
    |cx| passes!(cx),
    r#"
    let obj = { foo: 1 };
    obj.foo = 2;
    obj.foo = 3;
    obj.foo = 4;
    obj.foo = 5;
    invalidate();
    g = obj.foo;
"#
);

case!(
    switch_local,
    |cx| passes!(cx),
    r#"
    let a = { inner: 0, outer: 1 };
    switch (x) {
        case 1:
            a.inner = 2;
            a.inner;
        case 2:
            a.inner;
    }
    a.outer;
"#
);

case!(
    call_receiver,
    |cx| passes!(cx),
    r#"
    let something = { x: function() {} };
    something.x(); // receives `this`: do not opt
    let something2 = { x: function() {} };
    (0, something2.x)(); // does not receive `this`: opt
"#
);

case!(
    bail_bad_props1,
    |cx| passes!(cx),
    r#"
    let something = { __proto__: 1, x: 2 };
    g1 = something.x;
    something.x = g2;
"#
);

case!(
    bail_bad_props2,
    |cx| passes!(cx),
    r#"
    let something = { x: 2 };
    g1 = something.hasOwnProperty;
    something.x = g2;
"#
);

case!(
    bail_bad_props3,
    |cx| passes!(cx),
    r#"
    let something = { x: 2 };
    g1 = something.x;
    something.constructor = g2;
"#
);
