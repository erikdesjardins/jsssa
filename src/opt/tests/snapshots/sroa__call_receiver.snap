---
created: "2019-04-20T20:31:59.097807500Z"
creator: insta@0.7.4
source: src/opt/tests/sroa.rs
expression: "\n    let something = { x: function() {} };\n    something.x(); // receives `this`: do not opt\n    let something2 = { x: function() {} };\n    (0, something2.x)(); // does not receive `this`: opt\n"
---
_key = "x"
_val = <function>:
    <empty>
_ini = { [_key]: _val }
_prp = "x"
<dead> = _ini[_prp]()
<dead> = "x"
_val$1 = <function>:
    <empty>
x <= _val$1
<dead> = "x"
_fun = *x
<dead> = _fun()
